AQL query is not re-attempted on DatabaseException

The query should be retired after the following exception, but it wasn't.
```
Exception in thread "main" za.co.absa.spline.persistence.DatabaseException: Query execution failed: 
	at za.co.absa.spline.persistence.ArangoImplicits$ArangoDatabaseAsyncScalaWrapper$$anonfun$queryAs$1.applyOrElse(ArangoImplicits.scala:91)
	at za.co.absa.spline.persistence.ArangoImplicits$ArangoDatabaseAsyncScalaWrapper$$anonfun$queryAs$1.applyOrElse(ArangoImplicits.scala:88)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at scala.util.Failure.recover(Try.scala:234)
	at scala.concurrent.Future.$anonfun$recover$1(Future.scala:395)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
Caused by: com.arangodb.ArangoDBException: Response: 409, Error: 1210 - AQL: Error message received from cluster node 'PRMR-umszgjfq': unique constraint violated - in index idx_1729835903116902403 of type persistent over 'uri'; conflicting key: 186202774 (while executing)
	at com.arangodb.internal.util.ResponseUtils.checkError(ResponseUtils.java:55)
	at com.arangodb.internal.velocystream.VstCommunication.checkError(VstCommunication.java:157)
	at com.arangodb.async.internal.velocystream.VstCommunicationAsync.lambda$execute$1(VstCommunicationAsync.java:80)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at com.arangodb.async.internal.velocystream.VstConnectionAsync.lambda$write$0(VstConnectionAsync.java:51)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.arangodb.internal.velocystream.internal.MessageStore.consume(MessageStore.java:62)
	at com.arangodb.internal.velocystream.internal.ChunkStore.checkCompleteness(ChunkStore.java:65)
	at com.arangodb.internal.velocystream.internal.ChunkStore.checkCompleteness(ChunkStore.java:59)
	at com.arangodb.internal.velocystream.internal.VstConnection.lambda$open$0(VstConnection.java:218)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/commons/src/main/scala/za/co/absa/spline/common/AsyncCallRetryer.scala
+++ b/commons/src/main/scala/za/co/absa/spline/common/AsyncCallRetryer.scala
@@ -14,29 +14,20 @@
  * limitations under the License.
  */
 
-package za.co.absa.spline.persistence
+package za.co.absa.spline.common
 
-import java.util.concurrent.CompletionException
-
-import com.arangodb.ArangoDBException
 import org.slf4s.Logging
+import za.co.absa.spline.common.AsyncCallRetryer._
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
-object Persister extends Logging {
+class AsyncCallRetryer(isRetryable: Throwable => Boolean, maxRetries: Int) extends Logging {
 
-  import scala.concurrent.ExecutionContext.Implicits._
-
-  private[persistence] val MaxRetries = 5
-
-  def execute[R](fn: => Future[R]): Future[R] = {
+  def execute[R](fn: => Future[R])(implicit ex: ExecutionContext): Future[R] = {
     executeWithRetry(fn, None)
   }
 
-  @throws(classOf[IllegalArgumentException])
-  @throws(classOf[ArangoDBException])
-  @throws(classOf[CompletionException])
-  private def executeWithRetry[R](fn: => Future[R], lastFailure: Option[FailedAttempt]): Future[R] = {
+  private def executeWithRetry[R](fn: => Future[R], lastFailure: Option[FailedAttempt])(implicit ex: ExecutionContext): Future[R] = {
     val eventualResult = fn
     val attemptsUsed = lastFailure.map(_.count).getOrElse(0)
 
@@ -44,14 +35,15 @@ object Persister extends Logging {
       log.warn(s"Succeeded after ${failure.count + 1} attempts. Previous message was: ${failure.error.getMessage}")
     }
 
-    if (attemptsUsed >= MaxRetries)
+    if (attemptsUsed >= maxRetries)
       eventualResult
     else
       eventualResult.recoverWith {
-        case RetryableException(e) => executeWithRetry(fn, Some(FailedAttempt(attemptsUsed + 1, e)))
+        case e if isRetryable(e) => executeWithRetry(fn, Some(FailedAttempt(attemptsUsed + 1, e)))
       }
   }
+}
 
-  case class FailedAttempt(count: Int, error: Exception)
-
+object AsyncCallRetryer {
+  case class FailedAttempt(count: Int, error: Throwable)
 }

--- a/persistence/src/main/scala/za/co/absa/spline/persistence/RetryableExceptionUtils.scala
+++ b/persistence/src/main/scala/za/co/absa/spline/persistence/RetryableExceptionUtils.scala
@@ -16,12 +16,12 @@
 
 package za.co.absa.spline.persistence
 
-import java.util.concurrent.CompletionException
-
 import com.arangodb.ArangoDBException
 import za.co.absa.spline.persistence.ArangoCode._
 
-object RetryableException {
+import scala.annotation.tailrec
+
+object RetryableExceptionUtils {
 
   private[persistence] val RetryableCodes = Set(
     ArangoConflict,
@@ -33,10 +33,16 @@ object RetryableException {
     ClusterTimeout)
     .map(_.code)
 
-  def unapply(exception: Throwable): Option[RuntimeException] = exception match {
-    case e: ArangoDBException if RetryableCodes(e.getResponseCode) => Some(e)
-    case e: CompletionException => Option(e.getCause).flatMap(unapply).map(_ => e)
-    case _ => None
+  def isRetryable(exception: Throwable): Boolean = {
+    @tailrec def loop(ex: Throwable, visited: Set[Throwable]): Boolean = {
+      (!visited.contains(ex)) && (ex match {
+        case e: ArangoDBException if RetryableCodes.contains(e.getErrorNum) => true
+        case e: Exception => loop(e.getCause, visited + e)
+        case _ => false
+      })
+    }
+
+    loop(exception, Set.empty)
   }
 
 }

--- a/persistence/src/test/scala/za/co/absa/spline/persistence/AsyncCallRetryerSpec.scala
+++ b/persistence/src/test/scala/za/co/absa/spline/persistence/AsyncCallRetryerSpec.scala
@@ -16,34 +16,28 @@
 
 package za.co.absa.spline.persistence
 
-import java.util.concurrent.CompletionException
-
-import com.arangodb.ArangoDBException
 import org.mockito.Mockito._
 import org.scalatest.flatspec.AsyncFlatSpec
 import org.scalatest.matchers.should.Matchers
 import org.scalatestplus.mockito.MockitoSugar
-import za.co.absa.spline.persistence.PersisterSpec.ARecoverableArangoErrorCode
+import za.co.absa.spline.common.AsyncCallRetryer
 
 import scala.concurrent.Future
 import scala.concurrent.Future._
 import scala.language.implicitConversions
 
-object PersisterSpec {
-  private val ARecoverableArangoErrorCode = ArangoCode.ClusterTimeout.code
-}
-
-class PersisterSpec
+class AsyncCallRetryerSpec
   extends AsyncFlatSpec
     with Matchers
     with MockitoSugar {
 
-  behavior of "Persister"
+  behavior of "isRetryable()"
 
   it should "call an underlying method and return a result" in {
+    val retryer = new AsyncCallRetryer(mock[Throwable => Boolean], 5)
     val spy = mock[() => Future[String]]
     when(spy()) thenReturn successful("result")
-    for (result <- Persister.execute(spy()))
+    for (result <- retryer.execute(spy()))
       yield {
         verify(spy, times(1))()
         result should equal("result")
@@ -51,36 +45,35 @@ class PersisterSpec
   }
 
   it should "retry after a recoverable failure" in {
-    Future.traverse(RetryableException.RetryableCodes) {
-      errorCode => {
-        val spy = mock[() => Future[String]]
-        (when(spy())
-          thenReturn failed(new ArangoDBException("1st call failed", errorCode))
-          thenReturn failed(new CompletionException(new ArangoDBException("2st call failed", errorCode)))
-          thenReturn successful("3rd call succeeded"))
-        for (result <- Persister.execute(spy()))
-          yield {
-            verify(spy, times(3))()
-            result should equal("3rd call succeeded")
-          }
+    val retryer = new AsyncCallRetryer(_ => true, 5)
+    val spy = mock[() => Future[String]]
+    (when(spy())
+      thenReturn failed(new Exception("1st call failed"))
+      thenReturn failed(new Exception("2nd call failed"))
+      thenReturn successful("3rd call succeeded"))
+    for (result <- retryer.execute(spy()))
+      yield {
+        verify(spy, times(3))()
+        result should equal("3rd call succeeded")
       }
-    }.map(_.head)
   }
 
   it should "only retry up to the maximum number of retries" in {
+    val retryer = new AsyncCallRetryer(_ => true, 2)
     val spy = mock[() => Future[String]]
-    when(spy()) thenReturn failed(new ArangoDBException("oops", ARecoverableArangoErrorCode))
-    for (thrown <- recoverToExceptionIf[ArangoDBException](Persister.execute(spy())))
+    when(spy()) thenReturn failed(new Exception("oops"))
+    for (thrown <- recoverToExceptionIf[Exception](retryer.execute(spy())))
       yield {
-        verify(spy, times(Persister.MaxRetries + 1))()
+        verify(spy, times(3))() // 2 retries + 1 initial call
         thrown.getMessage should equal("oops")
       }
   }
 
   it should "not retry on a non-recoverable error" in {
+    val retryer = new AsyncCallRetryer(_ => false, 5)
     val spy = mock[() => Future[String]]
     when(spy()) thenReturn failed(new RuntimeException("boom"))
-    for (thrown <- recoverToExceptionIf[Exception](Persister.execute(spy())))
+    for (thrown <- recoverToExceptionIf[Exception](retryer.execute(spy())))
       yield {
         verify(spy, times(1))()
         thrown.getMessage should equal("boom")

--- a/persistence/src/test/scala/za/co/absa/spline/persistence/RetryableExceptionUtilsSpec.scala
+++ b/persistence/src/test/scala/za/co/absa/spline/persistence/RetryableExceptionUtilsSpec.scala
@@ -1,0 +1,109 @@
+/*
+ * Copyright 2022 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.persistence
+
+import com.arangodb.ArangoDBException
+import com.arangodb.entity.ErrorEntity
+import com.arangodb.velocypack.{VPack, VPackBuilder, ValueType}
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.{BeforeAndAfterEach, OptionValues}
+import za.co.absa.spline.persistence.RetryableExceptionUtils.RetryableCodes
+import za.co.absa.spline.persistence.RetryableExceptionUtilsSpec._
+
+class RetryableExceptionUtilsSpec
+  extends AnyFlatSpec
+    with BeforeAndAfterEach
+    with Matchers
+    with OptionValues {
+
+
+  behavior of "RetryableException.unapply()"
+
+  RetryableCodes.foreach { errNum =>
+    it should s"return Some(ex) when ex is ArangoDBException with error $errNum" in {
+      val ex =
+        new ArangoDBException(
+          new ErrorEntity().copy(errorNum = errNum)
+        )
+      RetryableExceptionUtils.isRetryable(ex) shouldBe true
+    }
+  }
+
+  it should s"return Some(ex) when ex is ArangoDBException wrapped with another exception" in {
+    val ex =
+      new Exception(
+        new RuntimeException(
+          new ArangoDBException(
+            new ErrorEntity().copy(errorNum = RetryableCodes.head)
+          )
+        )
+      )
+    RetryableExceptionUtils.isRetryable(ex) shouldBe true
+  }
+
+  it should "return None when neither ex nor any of its causes is ArangoDBException with retryable error codes" in {
+    RetryableExceptionUtils.isRetryable(new Exception()) shouldBe false
+    RetryableExceptionUtils.isRetryable(new ArangoDBException(new ErrorEntity().copy(errorNum = ArangoCode.Internal.code))) shouldBe false
+  }
+
+  it should "gracefully handle nulls" in {
+    RetryableExceptionUtils.isRetryable(null) shouldBe false
+  }
+
+  it should "gracefully handle exceptions with null cause" in {
+    RetryableExceptionUtils.isRetryable(new Exception) shouldBe false
+  }
+
+  it should "gracefully handle exceptions with looped cause" in {
+    lazy val loopedEx: Exception =
+      new Exception {
+        override def getCause: Throwable =
+          new Exception {
+            override def getCause: Throwable = loopedEx
+          }
+      }
+    RetryableExceptionUtils.isRetryable(loopedEx) shouldBe false
+  }
+
+}
+
+object RetryableExceptionUtilsSpec {
+
+  private val vpack = new VPack.Builder().build
+
+  implicit class ErrorEntityOps(val ee: ErrorEntity) {
+    def copy(
+      errorMessage: String = ee.getErrorMessage,
+      exception: String = ee.getException,
+      code: Int = ee.getCode,
+      errorNum: Int = ee.getErrorNum
+    ): ErrorEntity = {
+      vpack.deserialize[ErrorEntity](
+        new VPackBuilder()
+          .add(ValueType.OBJECT)
+          .add("errorMessage", errorMessage)
+          .add("exception", exception)
+          .add("code", Int.box(code))
+          .add("errorNum", Int.box(errorNum))
+          .close()
+          .slice(),
+        classOf[ErrorEntity]
+      )
+    }
+  }
+}

--- a/producer-services/src/main/scala/za/co/absa/spline/producer/service/ProducerServicesConfig.scala
+++ b/producer-services/src/main/scala/za/co/absa/spline/producer/service/ProducerServicesConfig.scala
@@ -16,8 +16,14 @@
 
 package za.co.absa.spline.producer.service
 
-import org.springframework.context.annotation.{ComponentScan, Configuration}
+import org.springframework.context.annotation.{Bean, ComponentScan, Configuration}
+import za.co.absa.spline.common.AsyncCallRetryer
+import za.co.absa.spline.persistence.RetryableExceptionUtils
 
 @Configuration
 @ComponentScan(basePackageClasses = Array(classOf[repo._package]))
-class ProducerServicesConfig
+class ProducerServicesConfig {
+
+  @Bean def repeater: AsyncCallRetryer = new AsyncCallRetryer(isRetryable = RetryableExceptionUtils.isRetryable, 5)
+
+}

--- a/producer-services/src/main/scala/za/co/absa/spline/producer/service/repo/ExecutionProducerRepositoryImpl.scala
+++ b/producer-services/src/main/scala/za/co/absa/spline/producer/service/repo/ExecutionProducerRepositoryImpl.scala
@@ -20,13 +20,14 @@ import com.arangodb.async.ArangoDatabaseAsync
 import org.slf4s.Logging
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.stereotype.Repository
+import za.co.absa.spline.common.AsyncCallRetryer
 import za.co.absa.spline.persistence.model._
 import za.co.absa.spline.persistence.tx.{ArangoTx, InsertQuery, TxBuilder}
-import za.co.absa.spline.persistence.{ArangoImplicits, Persister}
+import za.co.absa.spline.persistence.ArangoImplicits
 import za.co.absa.spline.producer.model.v1_1.ExecutionEvent._
 import za.co.absa.spline.producer.model.{v1_1 => apiModel}
 import za.co.absa.spline.producer.service.model.{ExecutionEventKeyCreator, ExecutionPlanPersistentModel, ExecutionPlanPersistentModelBuilder}
-import za.co.absa.spline.producer.service.{UUIDCollisionDetectedException, InconsistentEntityException}
+import za.co.absa.spline.producer.service.{InconsistentEntityException, UUIDCollisionDetectedException}
 
 import java.util.UUID
 import scala.compat.java8.FutureConverters._
@@ -35,13 +36,13 @@ import scala.concurrent.{ExecutionContext, Future}
 import scala.util.control.NonFatal
 
 @Repository
-class ExecutionProducerRepositoryImpl @Autowired()(db: ArangoDatabaseAsync) extends ExecutionProducerRepository
+class ExecutionProducerRepositoryImpl @Autowired()(db: ArangoDatabaseAsync, repeater: AsyncCallRetryer) extends ExecutionProducerRepository
   with Logging {
 
   import ArangoImplicits._
   import ExecutionProducerRepositoryImpl._
 
-  override def insertExecutionPlan(executionPlan: apiModel.ExecutionPlan)(implicit ec: ExecutionContext): Future[Unit] = Persister.execute({
+  override def insertExecutionPlan(executionPlan: apiModel.ExecutionPlan)(implicit ec: ExecutionContext): Future[Unit] = repeater.execute({
     // Here I have to use the type parameter `Any` and cast to `String` later due to ArangoDb Java driver issue.
     // See https://github.com/arangodb/arangodb-java-driver/issues/389
     val eventualMaybeExistingDiscriminatorOpt: Future[Option[String]] = db.queryOptional[Any](
@@ -80,7 +81,7 @@ class ExecutionProducerRepositoryImpl @Autowired()(db: ArangoDatabaseAsync) exte
     } yield Unit
   })
 
-  override def insertExecutionEvents(events: Array[apiModel.ExecutionEvent])(implicit ec: ExecutionContext): Future[Unit] = Persister.execute({
+  override def insertExecutionEvents(events: Array[apiModel.ExecutionEvent])(implicit ec: ExecutionContext): Future[Unit] = repeater.execute({
     val eventualExecPlanInfos: Future[Seq[ExecPlanInfo]] = db.queryStream[ExecPlanInfo](
       s"""
          |WITH executionPlan, executes, operation, dataSource

